File: prince-of-versions/src/main/java/co/infinum/princeofversions/PrinceOfVersionsDefaultExecutor.java
package co.infinum.princeofversions;

import java.util.concurrent.Executor;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.ThreadFactory;

import javax.annotation.Nonnull;

/**
 * Implementation of {@link Executor} which runs check on background thread of default priority.
 */
final class PrinceOfVersionsDefaultExecutor implements Executor {

    private static final ExecutorService SERVICE = Executors.newSingleThreadExecutor(new ThreadFactory() {
        @Override
        public Thread newThread(@Nonnull Runnable r) {
            Thread t = new Thread(r, "PrinceOfVersions Thread");
            t.setDaemon(true);
            return t;
        }
    });

    @Override
    public void execute(@Nonnull Runnable runnable) {
        SERVICE.submit(runnable);
    }
}

File: prince-of-versions/src/main/java/co/infinum/princeofversions/PrinceOfVersionsDefaultExecutor.kt
package co.infinum.princeofversions

import java.util.concurrent.Executor
import java.util.concurrent.Executors

/**
 * An implementation of [Executor] that runs the update check on a single background thread.
 */
internal class PrinceOfVersionsDefaultExecutor : Executor {

    private val service = Executors.newSingleThreadExecutor { runnable ->
        Thread(runnable, "PrinceOfVersions Thread").apply {
            // Set as a daemon thread so it doesn't prevent the JVM from exiting.
            isDaemon = true
        }
    }

    override fun execute(runnable: Runnable) {
        service.submit(runnable)
    }
}
